src/datafusion/helpers.rs

```diff
@@ -0,0 +1,228 @@
+        | Expr::SimilarTo { .. }
+        | Expr::InList { .. }
+        | Expr::Exists { .. }
+        | Expr::InSubquery(_)
+        | Expr::ScalarSubquery(_)
+        | Expr::GetIndexedField { .. }
+        | Expr::GroupingSet(_)
+        | Expr::Case { .. } => Ok(VisitRecursion::Continue),
+
+        Expr::ScalarFunction(scalar_function) => match &scalar_function.func_def {
+            ScalarFunctionDefinition::BuiltIn(fun) => match fun.volatility() {
+                Volatility::Immutable => Ok(VisitRecursion::Continue),
+                Volatility::Stable | Volatility::Volatile => {
+                    is_applicable = false;
+                    Ok(VisitRecursion::Stop)
+                }
+            },
+            ScalarFunctionDefinition::UDF(fun) => match fun.signature().volatility {
+                Volatility::Immutable => Ok(VisitRecursion::Continue),
+                Volatility::Stable | Volatility::Volatile => {
+                    is_applicable = false;
+                    Ok(VisitRecursion::Stop)
+                }
+            },
+            ScalarFunctionDefinition::Name(_) => {
+                internal_err!("Function `Expr` with name should be resolved.")
+            }
+        },
+
+        Expr::AggregateFunction { .. }
+        | Expr::Sort { .. }
+        | Expr::WindowFunction { .. }
+        | Expr::Wildcard { .. }
+        | Expr::Unnest { .. }
+        | Expr::Placeholder(_) => {
+            is_applicable = false;
+            Ok(VisitRecursion::Stop)
+        }
+    })
+    .unwrap();
+    is_applicable
+}
+
+// Below is all the logic necessary (I think) to convert a PhysicalExpr into a ZarrChunkFilter.
+// The logic is mostly copied from datafusion, and is simplified here for the zarr use case.
+pub struct ZarrFilterCandidate {
+    expr: Arc<dyn PhysicalExpr>,
+    projection: Vec<usize>,
+}
+
+struct ZarrFilterCandidateBuilder<'a> {
+    expr: Arc<dyn PhysicalExpr>,
+    file_schema: &'a Schema,
+    required_column_indices: BTreeSet<usize>,
+}
+
+impl<'a> ZarrFilterCandidateBuilder<'a> {
+    pub fn new(expr: Arc<dyn PhysicalExpr>, file_schema: &'a Schema) -> Self {
+        Self {
+            expr,
+            file_schema,
+            required_column_indices: BTreeSet::default(),
+        }
+    }
+
+    pub fn build(mut self) -> DataFusionResult<Option<ZarrFilterCandidate>> {
+        let expr = self.expr.clone().rewrite(&mut self)?;
+
+        Ok(Some(ZarrFilterCandidate {
+            expr,
+            projection: self.required_column_indices.into_iter().collect(),
+        }))
+    }
+}
+
+impl<'a> TreeNodeRewriter for ZarrFilterCandidateBuilder<'a> {
+    type N = Arc<dyn PhysicalExpr>;
+
+    fn pre_visit(&mut self, node: &Arc<dyn PhysicalExpr>) -> DataFusionResult<RewriteRecursion> {
+        if let Some(column) = node.as_any().downcast_ref::<Column>() {
+            if let Ok(idx) = self.file_schema.index_of(column.name()) {
+                self.required_column_indices.insert(idx);
```

Comment:
So first, we accumulate indices of columns required (by a given predicate). These indices represent the position of the column in the file schema (which will eventually be the table schema, for now we don't have that distinction), e.g. if the predicate requires `lat, lon` and the file schema is `float_data, lat, lon`, we will end up setting the projection to `[1, 2]`. Since the set is ordered, I think even if in the predicate the order was `lon, lat`, we'd end up with `[1, 2]` as the projection.